module/VuFind/src/VuFind/I18n/Translator/TranslatorAwareTrait.php

```diff
@@ -114,7 +117,17 @@ public function translate($target, $tokens = [], $default = null)
                 return $translated;
             }
             // Override $domain/$str using getDisplayString() before proceeding:
-            list($domain, $str) = $this->extractTextDomain($str->getDisplayString());
+            $str = $str->getDisplayString();
+            // Also the display string can be a TranslatableString. This makes it
+            // possible have multiple levels of translatable values while still
+            // providing a sane default string if translation is not found. Used at
+            // least with hierarchical facets where translation key can be the exact
+            // facet value (e.g. "0/Book/") or a displayable value (e.g. "Book").
+            if ($str instanceof \VuFind\I18n\TranslatableStringInterface) {
+                $str = $this->translate($str);
```

Comment:
Is it okay to keep the existing value of `$domain` that was extracted earlier in this situation, or do we need to reset the domain in this case? (I'm guessing that this makes no difference most of the time... but just trying to think whether there are any edge cases where it could lead to unexpected behavior).

Reply:
I think so, since translate() will handle any domain and return a domain-less translation result. There's still the else branch below for the existing behavior.

Reply:
I was able to put together a test cause that caused this to break -- if I put a translation from "failure" to "success" in the initial text domain, and then default to a string of just "failure" without a text domain, it gets incorrectly translated to "success." See fea4a89aecb07fd8cbce7960ee941815ef6e670f -- I believe the solution is that we need to immediately recurse in this edge case rather than dropping through to the translateString() call. I'm going to do a bit more testing, and will merge the PR if I see no further problems... but I'd appreciate it if you could double-check my fix in case you have any remaining concerns.